gradio/i18n.py

```diff
@@ -0,0 +1,130 @@
+        """
+        self.key = key
+        self._type = "translation_metadata"
+
+    def to_dict(self) -> dict[str, Any]:
+        """
+        Convert the TranslationMetadata object to a dictionary for serialization.
+        This allows the frontend to recognize it as a translatable object.
+        """
+        return {"__type__": self._type, "key": self.key}
+
+    def __str__(self) -> str:
+        """
+        String representation of the TranslationMetadata object.
+        Used when the object is converted to a string.
+        This returns a special format that can be recognized by the frontend
+        as needing translation.
+        """
+        import json
+
+        # Create a JSON representation that can be detected and processed by the frontend
+        return f"__i18n__{json.dumps(self.to_dict())}"
+
+    def __repr__(self) -> str:
+        """
+        Representation of the TranslationMetadata object for debugging.
+        """
+        return self.__str__()
+
+    # Add more string compatibility methods
+    def __add__(self, other):
+        """
+        Handle string concatenation (self + other).
+        """
+        return str(self) + str(other)
+
+    def __radd__(self, other):
+        """
+        Handle string concatenation (other + self).
+        """
+        return str(other) + str(self)
+
+    # For attribute/method access attempts to be gracefully handled
+    def __getattr__(self, name):
+        """
+        Handle attribute access for TranslationMetadata.
+        This makes it possible to use TranslationMetadata objects in contexts
+        that expect strings with methods.
+        """
+        if name.startswith("__") and name.endswith("__"):
+            raise AttributeError(f"{self.__class__.__name__} has no attribute {name}")
+
+        # Return a function that returns self for any method call
+        def method(*_args, **_kwargs):
+            return self
+
+        return method
+
+    # Make the object JSON-serializable
+    def tojson(self) -> dict[str, Any]:
+        """
+        Convert the TranslationMetadata object to a JSON-serializable dictionary.
+        This is used by the default Python JSON serializer.
+        """
+        return self.to_dict()
+
+
+class I18n:
+    """
+    Handles internationalization (i18n) for Gradio applications.
+
+    Stores translation dictionaries and provides a method to retrieve translation keys.
+    The translation lookup happens on the frontend based on the browser's locale
+    and the provided translation dictionaries.
+    """
+
+    def __init__(self, **translations: dict[str, str]):
+        """
+        Initializes the I18n class.
+
+        Args:
+            **translations: Each keyword argument should be a locale code (e.g., "en", "fr") with a
+                          dictionary value, which maps translation keys to translated strings.
+                          Example: gr.I18n(en={"greeting": "Hello"}, es={"greeting": "Hola"})
```

Comment:
Is there an easy way to detect invalid locales so that this doesn't fail silently?